### PR TITLE
Abstracts joel (5, 9-14)

### PR DESCRIPTION
**Pull request by joelostblom**
Requesting review from @DamienIrving  or @gvwilson  since I believe you have written most of the material in many of these chapters.

For most of the abstracts, I melded the chapter intro with the keypoints (thanks for writing these so well), while glancing at the chapter headings to see that most of them were included. They became a bit longer than intended, but are within the 100-200 word range required. There is also some very minor cleanup of the files I looked at.

Tagging @cwickham in case this workflow tip is helpful for you when writing your abstracts.

### CODE REVIEW

File: chapters/chapter-abstracts.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
+## 5. Building Command-Line Tools with Python
+
+Writing command-line Python scripts allows us to run our programs from the Unix shell,
+which facilitates both scalability and interoperability with data analysis pipelines.
+The `argparse` Python module can be used to parse user input,
+and provides other expected command line functionality
+such as positional and optional arguments, and help messages.
+Python scripts can take input from files or directly from the standards input
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
Python scripts can take input from files or directly from standard input
```

File: chapters/chapter-abstracts.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+and write to fies or the standard output.
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md:9]:
```suggestion
and write to files or standard output.
```

File: chapters/chapter-abstracts.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+When we write Python scripts we put the put `import` statements first,
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
When we write Python scripts we put the `import` statements first,
```

File: chapters/chapter-abstracts.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+according to functionality
+and to document the behavior of each module and function carefully
+in their respective docstring.
+
+## 9. Automating Analyses with Make
+
+Keeping track of programs that process multiple data files
+is an error prone process that we should automate to minimize mistakes.
+Build-managers such as `make`
+can keep track of file dependencies in the analysis pipelines
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md:25]:
```suggestion
can keep track of file dependencies in analysis pipelines
```

File: chapters/chapter-abstracts.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+it is recommended to modularize code into different files 
+according to functionality
+and to document the behavior of each module and function carefully
+in their respective docstring.
+
+## 9. Automating Analyses with Make
+
+Keeping track of programs that process multiple data files
+is an error prone process that we should automate to minimize mistakes.
+Build-managers such as `make`
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
Build-managers such as Make
```
During the chapter we refer to Make as opposed to `make`

File: chapters/chapter-abstracts.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+A target can be a file or a phony target that triggers an action.
+When a target is out of date with respect to its prerequisites,
+Make executes the recipe associated with its rule.
+Make defines a few convenience variables
+such as `$@` (target),
+`$^` (all prerequisites),
+and `$<` (first prerequisite).
+Makefiles can also use wildcard matching of file names
+and functions such as `$(patsubst ...)` for increased flexibility.
+As always it is important to document our code
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
As always, it is important to document our code
```

File: chapters/chapter-abstracts.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+and in Make we use specially-formatted comments to create self-documenting Makefiles.
+
+## 10. Configuring Programs
+
+For complex programs
+it can improve usability and increase reproducibility
+to be able to read configuration options from files
+in addition to using command-line options.
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md:46]: Perhaps this sentence would read better as the following?
"For complex programs, usability and reproducibility can be improved by reading configuration options from files in addition to using command-line options."

File: chapters/chapter-abstracts.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+We may want to use up to four layers of configuration:
+a system-wide configuration file for general settings;
+a user-specific configuration file for personal preferences;
+a job-specific file with settings for a particular run;
+and command-line options to change things that commonly change.
+This is sometimes called overlay configuration
+because each level overrides the ones above it.
+Often,
+a pre-existing standard such as the YAML or INI syntax
+is used to write these configuration files.
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
is used to write configuration files.
```

File: chapters/chapter-abstracts.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+Saving and sharing these configuration files
+makes it easy for researchers programs with the same settings,
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
makes it easy for researchers to run programs with the same settings,
```

File: chapters/chapter-abstracts.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+which reduces the risk for errors and increases reproducibility.
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
which reduces the risk of errors and increases reproducibility.
```

File: chapters/chapter-abstracts.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+## 11. Testing Software
+
+It is critical to test software to reassure people (including yourself)
+that the software is correct enough
+and to make tolerances on "enough" explicit.
+There are several layers to testing software:
+adding assertions to code so that it checks itself as it runs;
+writing unit tests to check individual pieces of code;
+writing integration tests to check that those pieces work together correctly;
+writing regression tests to check if things that used to work no longer do.
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
and writing regression tests to check if things that used to work no longer do.
```

File: chapters/chapter-abstracts.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+Fortunately,
+there are many excellent test frameworks to aid us with these tasks.
+A test framework finds and runs tests written in a prescribed fashion
+and reports their results.
+In addition to test failures,
+these reports often include the tests' coverage,
+which is the fraction of lines of code that are checked.
+To automate the execution of our tests
+we can use Continuous Integration
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
we can use a continuous integration tool like Travis CI,
```

File: chapters/chapter-abstracts.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+Generally speaking,
+we distinguish between two types of errors/exceptions.
+Internal errors are mistakes in the program itself,
+such as calling a function with None instead of a list.
+External errors are usually caused by interactions
+between the program and the outside world:
+a user may mis-type a filename,
+the network might be down,
+and so on.
+
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md:108]: This handling errors abstract is incomplete. It needs to talk about all the things we do in this chapter, like raising and catching exceptions, logging, etc.

File: chapters/chapter-abstracts.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+## 13. Tracking Provenance
+
+Modern publishing involves much more than producing a PDF
+and making it available on a preprint server.
+It also entails providing the data underpinning the report
+and the code used to do the analysis.
+While some reports, datasets, software packages, and analysis scripts
+can't be published without violating personal or commercial confidentiality,
+every researcher's default should be to make all these as widely available as possible.
+Publishing it under an open license is the first step.
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
Publishing under an open license is the first step.
```

File: chapters/chapter-abstracts.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+Using DOIs facilitates identification of reports, datasets, and software releases.
+Similarly,
+using an ORCID ensures that your research profile it correctly linked with your work
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
using an ORCID ensures that your research profile is correctly linked with your work
```

File: chapters/chapter-abstracts.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+and not mixed up with others of the same name.
+Any published data should be FAIR:
+findable, accessible, interoperable, and reusable.
+While small datasets can be put under standard version control;
+large ones should be stored on data sharing sites,
+sometimes there are specific ones for you field or the technology used for data collection.
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
Big datasets may not be ours in the first place, and probably need the attention of a professional archivist.
```

File: chapters/chapter-abstracts.md [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+every researcher's default should be to make all these as widely available as possible.
+Publishing it under an open license is the first step.
+Using DOIs facilitates identification of reports, datasets, and software releases.
+Similarly,
+using an ORCID ensures that your research profile it correctly linked with your work
+and not mixed up with others of the same name.
+Any published data should be FAIR:
+findable, accessible, interoperable, and reusable.
+While small datasets can be put under standard version control;
+large ones should be stored on data sharing sites,
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
medium-sized ones should be stored on data sharing sites like Figshare or Dryad.
```

File: chapters/chapter-abstracts.md [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+can't be published without violating personal or commercial confidentiality,
+every researcher's default should be to make all these as widely available as possible.
+Publishing it under an open license is the first step.
+Using DOIs facilitates identification of reports, datasets, and software releases.
+Similarly,
+using an ORCID ensures that your research profile it correctly linked with your work
+and not mixed up with others of the same name.
+Any published data should be FAIR:
+findable, accessible, interoperable, and reusable.
+While small datasets can be put under standard version control;
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
While small datasets can be put under standard version control,
```

File: chapters/chapter-abstracts.md [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+large ones should be stored on data sharing sites,
+sometimes there are specific ones for you field or the technology used for data collection.
+To facilitate for others to use your software, it is important that you
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
In terms of providing the code associated with a report, it is important that you
```

File: chapters/chapter-abstracts.md [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+describe your software environment, analysis scripts, and data processing steps
+in reproducible and easily inspectable ways.
+
+## 14. Creating Packages with Python
+
+While the core of programming languages are useful on their own,
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
While the libraries that come standard with a programming language are useful on their own,
```

File: chapters/chapter-abstracts.md [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+much of a language's value is added by community-contributed packages
+with domain-specific functionality.
+The most popular way to build packages in Python,
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
The most popular way to build packages in Python
```

File: chapters/chapter-abstracts.md [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+containing the package's source files.
+To track and communicate new functionality
+that we add to the package
+it is recommended to use semantic versioning.
+Before distributing our packages
+we want to ensure that they install in a fresh environment;
+for this we can use virtual Python environments
+instead having to worry about messing up our main Python installation.
+To ensure that others can use our package as we intend,
+it is instrumental that we write documentation in the package's README file
```
Comment (DamienIrving) [inline chapters/chapter-abstracts.md]:
```suggestion
it is customary to write documentation in the package's README file
```

**Review by DamienIrving — CHANGES_REQUESTED** [review]
Thanks for drafting these abstracts, @joelostblom.

I've made a bunch of minor suggestions in code comments - the only major comment I have is that [the instructions](https://github.com/merely-useful/py-rse/blob/book/taylor-francis/abstracts-orcids.pdf) ask for a third person voice, which isn't consistently followed throughout the abstracts. Not sure how strict they will be on that?

Also, I guess we'll just tell the publishers that we don't have specific author lists for each chapter?

**Comment by gvwilson** [comment]
@DamienIrving has caught everything I would have said and more :-)

**Comment by joelostblom** [comment]
Thanks @DamienIrving and sorry for poor spellchecking. I will rewrite by EOD Wed.

File: chapters/chapter-abstracts.md (hunk 20 above)
Comment (gvwilson) [reply chapters/chapter-abstracts.md]: I would remove the word 'standard' from Damien's suggested fix.

**Comment by joelostblom** [comment]
@DamienIrving @gvwilson I believe I have addressed your comments now

**Pull request merged**